pyre/pyre-interpreter/src/module/_typing/_typing_app.py

```
@@ -529,20 +690,108 @@ def __init_subclass__(cls, **kwargs):
+                type(param) is TypeVar
+                or type(param) is ParamSpec
+                or type(param) is TypeVarTuple
+            ):
+                raise TypeError(f"Expected a type param, got {param!r}")
+            default = param.__default__
+            if default is NoDefault:
+                if default_seen:
+                    raise TypeError(
+                        f"non-default type parameter '{param!r}' "
+                        "follows default type parameter"
+                    )
```

Ruff (0.16.1)

[warning] Avoid specifying long messages outside the exception class

(TRY003)